more accurate file language detection using file content heuristics

sqs opened the pull request:
- use src-d's enry (Go linguist port) for lang detection
- clean up lang detection and inventory code
- use more accurate heuristics for lang detection

With this PR, we are now using enry (https://github.com/src-d/enry), src-d's Go port of GitHub's linguist, which detects the programming language of a file. Previously we only had it detect the language based on the filename. This leads to incorrect results for many extensions, such as `.m` (Objective-C, Mathematica, or Matlab?).

This commit adds a feature flag env var USE_ENHANCED_LANGUAGE_DETECTION (default true in dev, default false otherwise) that performs enhanced language detection using file contents. This is significantly slower because it needs to read most files from each repository.  The existing (default) behavior is virtually unchanged except that vendor files are now excluded.

We will evaluate the performance of the USE_ENHANCED_LANGUAGE_DETECTION mode and tweak it while the feature flag defaults to off. We intend to support accurate language detection and intend to eventually enable the feature flag by default.

(See commit messages for details.)

This is a precursor to better language detection and stats as described in #5287  #2586  #2587 #1235 and for https://app.hubspot.com/contacts/2762526/company/556068698 https://app.hubspot.com/contacts/2762526/company/464956351 and others.

keegancsmith commented:
> USE_ENHANCED_LANGUAGE_DETECTION

Should we maybe make this part of the config, so its easier to experiment with at customers requesting this functionality?

keegancsmith commented on cmd/frontend/backend/repos.go:
```diff
@@ -226,5 +242,7 @@ func (s *repos) GetInventoryUncached(ctx context.Context, repo *types.Repo, comm
 	}
-	return inventory.Get(ctx, files)
+	return inventory.Get(ctx, files, func(ctx context.Context, path string, maxBytes int64) ([]byte, error) {
+		return git.ReadFile(ctx, *cachedRepo, commitID, path, maxBytes)
```
I think this would possibly result in too much data, but I had a thought around more granular caching to improve this dramatically:

Cache key is the object sha. So we cache the calculation on both the tree and blobs. Alternatively just cache on the tree (should be signficantly less tress than blobs). The amount of cache shared between commits would be awesome.

sqs replied:
Ah, great idea to cache on the OID of the tree. I will implement that. :tada:

sqs replied:
Posting some benchmarks here before I implement this better caching:

- The old (existing) repo inventory for sourcegraph/sourcegraph takes ~500msec.
- The enhanced repo inventory for sourcegraph/sourcegraph takes ~2400msec. This is slower because it looks at file contents.

These numbers are on localhost and likely understate the perf hit of the enhanced repo inventory impl because gitserver and frontend are on the same host. So better caching definitely seems important.

keegancsmith commented on cmd/frontend/graphqlbackend/search.go:
```diff
@@ -778,29 +778,15 @@ func sortSearchSuggestions(s []*searchSuggestionResolver) {
 			}
-			extPatterns := make([]string, len(lang.Extensions))
-			for i, ext := range lang.Extensions {
+			exts := enry.GetLanguageExtensions(lang)
```
how is the perf of enry vs the thing we wrote in this case? I'm hoping better since we used regexes pretty naively. But in the past we have done perf optimizations on this since it showed up in slow requests.

sqs replied:
I'm glad you pointed this out. I just saw in the benchmark that `enry.IsVendor` is super super slow, so I need to fix that before merging.

Other than that, using file content for detection takes about 10x as long (~10ms per file) as what we had before, but we only need to run that on fewer files (due to eliminating vendor files, and skipping file content detection when the file extension is unambiguous), so I think it will be OK. That is one of the things I'd like to test once merged.

keegancsmith replied:
Ouch 10ms is quite slow if its per file. Is that CPU bound on enry, or including latencies of fetching files from gitserver? I can imagine quite a long wait on something like the kubernetes repo in that case.

sqs replied:
This is CPU-bound and does not include file fetching latency. Let me get exact perf numbers on a better benchmark and I’ll post back here.

sqs replied:
I was completely wrong (not sure what I was looking at). For prom-repo-tree.txt (1880 files), it takes 122msec total (for all 1880 files), or 65usec per file. So this is acceptable. In practice this will differ based on the repository because some repositories will have many files whose language can be determined unambiguously just by looking at the file extension, and some will have many files whose contents must be inspected to determine the language. In the benchmark corpus, about 10% of the files are ambiguous and require file content inspection, and they are only slightly slower (~75usec/file).

keegancsmith replied:
phew! thanks for the update :)

keegancsmith commented on cmd/frontend/graphqlbackend/search_results_test.go:
```diff
@@ -537,7 +537,7 @@ func TestSearchResolver_DynamicFilters(t *testing.T) {
 			expectedDynamicFilterStrs: map[string]struct{}{
 				`repo:^testRepo$`: {},
-				`lang:typescript`: {},
+				`lang:tsx`:        {},
```
could you be breaking existing saved searches here?

sqs replied:
Yes, good point. I will investigate.

sqs replied:
Fixed; now the old and better behavior prevails

keegancsmith commented on cmd/frontend/internal/inventory/inventory.go:
```diff
@@ -31,87 +26,55 @@ type Lang struct {
-func Get(ctx context.Context, files []os.FileInfo) (*Inventory, error) {
+// Get performs an inventory of the files passed in. If readFile is provided, the language detection
+// uses heuristics based on the file content for greater accuracy.
+func Get(ctx context.Context, files []os.FileInfo, readFile func(ctx context.Context, path string, maxBytes int64) ([]byte, error)) (*Inventory, error) {
```
:+1: to passing in readFile as a closure.

keegancsmith commented on cmd/frontend/internal/inventory/inventory.go:
```diff
@@ -31,87 +26,55 @@ type Lang struct {
-		// filelang.IsVendored function is slow (benchmark goes from
-		// 160ms to 0.5ms without the check). Currently Inventory is
-		// just used to determine which languages are in a repo, the
-		// relative usage (TotalBytes) is not exposed or used. So
```
is enry's IsVendor function faster?

sqs replied:
No, I need to find a way to fix this. Thanks for flagging.

sqs replied:
For the prom-repo-tree.txt test case with 1880 filenames, it takes 124msec total, or 66usec per filename. So it should be fine. This is significantly faster than the old IsVendor.

keegancsmith commented:
This could even become a feature, since we could show the inventory for sub directories. Really useful for monorepos.

sqs commented:
@keegancsmith 	PTAL at the last commit with per-tree caching and perf data!

sqs commented on cmd/frontend/internal/inventory/tree.go:
```diff
@@ -0,0 +1,86 @@
+}
+
+// Tree computes the inventory of languages for a tree. It caches the inventories of subtrees.
+func (c *Context) Tree(ctx context.Context, tree os.FileInfo) (inv *Inventory, err error) {
```
@keegancsmith 	This is the main impl of the caching tree traverser.

keegancsmith commented on cmd/frontend/internal/inventory/tree.go:
```diff
@@ -0,0 +1,86 @@
+	for lang, totalBytes := range langTotalBytes {
+		sum.Languages = append(sum.Languages, &Lang{Name: lang, TotalBytes: totalBytes})
+	}
+	sort.SliceStable(sum.Languages, func(i, j int) bool {
```
Normal sort is more appropriate here, since in this case it will have the same outcome. There shouldn't be any things which have the same comparison (and your input list is random since it involves iterating over a map).

keegancsmith commented on cmd/frontend/internal/inventory/tree.go:
```diff
@@ -0,0 +1,86 @@
+}
+
+func sum(langTotalBytes map[string]uint64) *Inventory {
+	var sum Inventory
```
minor: we know the size of sum.Languages. I would suggest allocating it.

keegancsmith commented on cmd/frontend/backend/repos.go:
```diff
@@ -183,48 +189,59 @@ func (s *repos) GetInventory(ctx context.Context, repo *types.Repo, commitID api
+			return nil
+		},
+		CacheSet: func(tree os.FileInfo, inv inventory.Inventory) {
+			b, err := json.Marshal(inv)
```
`&inv`?

keegancsmith commented on cmd/frontend/backend/repos.go:
```diff
@@ -183,48 +189,59 @@ func (s *repos) GetInventory(ctx context.Context, repo *types.Repo, commitID api
+		},
+	}
+
+	if !useEnhancedLanguageDetection && false {
```
gated to always be false?

sqs replied:
Whoops, I left this in for benchmarking.

sqs commented:
> Impl looks good to me, some comments inline.

Thank you! Addressed all points.

> How does it feel in practice as you move around commits in a large repo?

There's nothing in the UI that lets me "feel" the perf, but the benchmarks in the 2nd-to-last (as of now) commit message give a good idea. Another one I just ran: given that 1 commit in kubernetes/kubernetes has its inventory cached, the previous 1000 ancestor commits take 43ms each (on average) to compute their inventory.

keegancsmith commented:
> given that 1 commit in kubernetes/kubernetes has its inventory cached, the previous 1000 ancestor commits take 43ms each (on average) to compute their inventory.

Woah!! Very nice :) What a fun little problem.

sqs commented:
Yeah, thanks for the great idea for this kind of caching!
